_posts/2019-07-02-cbo-introduction.md

```diff
@@ -0,0 +1,285 @@
+on several factors, including shape of the query, filters and table statistics.
+I would like to tell you more about what the table statistics are in Presto and
+what information can be derived from them.
+
+# Background
+
+Before diving deep into how Presto analyzes statistics, let’s set up a stage so
+that our considerations are framed in some context. Let’s consider a Data
+Scientist who wants to know which customers spend most dollars with the
+company, based on history of orders (probably to offer them some discounts).
+They would probably fire up a query like this:
+
+```sql
+SELECT c.custkey, sum(l.price)
+FROM customer c, orders o, lineitem l
+WHERE c.custkey = o.custkey AND l.orderkey = o.orderkey
+GROUP BY c.custkey ORDER BY sum(l.price) DESC;
+```
+
+Now, Presto needs to create an execution plan for this query. It does so by
+first transforming a query to a plan in the simplest possible way — here it
+will create CROSS JOINS for `FROM customer c, orders o, lineitem l` part of the
+query and FILTER for `WHERE c.custkey = o.custkey AND l.orderkey = o.orderkey`.
+The initial plan is very naïve — CROSS JOINS will produce humongous amounts of
+intermediate data. There is no point in even trying to execute such a plan and
+Presto won’t do that. Instead, it applies transformation to make the plan more
+what user probably wanted, as shown below. Note: for succinctness, only part of
+the query plan is drawn, without aggregation (`GROUP BY`) and sorting (`ORDER
+BY`).
+
+![](/assets/blog/cbo-introduction/presto-eliminate-cross-join.png)
+
+Indeed, this is much better than the CROSS JOINS. But we can do even better, if
+we consider _cost_.
+
+# Cost-Based Optimizer
+
+Without going into database internals on how JOIN is implemented, let’s take
+for granted that it makes a big difference which table is right and which is
+left in the JOIN. (Simple explanation would be that the table on the right
+basically needs to be kept in the memory while JOIN result is calculated).
+Because of that, the following plans produce same result, but may have
+different execution time or memory requirements.
+
+![](/assets/blog/cbo-introduction/presto-join-flip.png)
+
+CPU time, memory requirements and network bandwidth usage are the three
+dimensions that contribute to query execution time, both in single query and
+concurrent workloads. These dimensions are captured as the _cost_ in Presto.
+
+Our Data Scientist knows that most of the customers made at least one order and
+every order had at least one item (and many orders had many items), so
+`lineitem` is the biggest table, `orders` is medium and `customer` is the
+smallest. When joining `customer` and `orders`, having `orders` on the right
+side of the JOIN is not a good idea! However, how the planner can know that? In
+the real world, the query planner cannot reliably deduce information just from
+table names. This is where table statistics kick in.
+
+## Table statistics
```

Comment:
This renders as `h2` and single `#` renders as `h1`, correctly.
However, on the rendered page, h1 has font-size 1.5em (24px) and h2 has a bigger font-size, 32px.

Reply:
@electrum would you like to have a look here?

Reply:
I fixed this. See `Define h1 font-size for blog posts` commit.